### PR TITLE
【Hackathon 6th No.29】为 paddle.nn.functional.max_unpool1d / max_unpool2d / max_unpool3d/paddle.nn.functional.kl_div 进行功能增强

### PR DESCRIPTION
**Pull request by NKNaN**
### PR types
Others

### PR changes
APIs

### Description
1. paddle.nn.functional.max_unpool1d / max_unpool2d / max_unpool3d 增加支持 int64 输入
2. paddle.nn.functional.max_unpool1d / max_unpool2d / max_unpool3d 的 output_size 参数的判断有 bug，输入正确的 output_size 会报错：
这个目前按 x -> out, indices = pool(x, ...) -> unpool(out, indices, ..., output_size=x.shape) 这样来检验的话有问题的是paddle.nn.functional.max_unpool1d，output_size 参数处理之前少加了一维，是否还存在其他问题？

3. paddle.nn.functional.kl_div 增加参数 log_target

### CODE REVIEW

File: paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h [hunk 1]
```diff
@@ -23,13 +23,19 @@ namespace phi {
 using Array1 = Eigen::DSizes<int64_t, 1>;
 template <typename T>
 struct KLDivLossBackward {
-  HOSTDEVICE KLDivLossBackward() {}
+  bool LogTarget = false;
+
+  HOSTDEVICE KLDivLossBackward(bool _log_target) : LogTarget(_log_target) {}
```
Comment (zhwesky2010) [inline paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h:28]: 形参应该是不带下划线的，结构体 内部变量你可以带个 下划线

File: paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h [hunk 2]
```diff
@@ -58,9 +65,9 @@ void KLDivLossGradKernel(const Context& dev_ctx,
   auto loss_grad_t = phi::EigenVector<T>::Flatten(*loss_grad);
 
   auto loss_grad_expand = loss_grad_t.broadcast(Array1(expand));
-  auto grad_t = target_t * loss_grad_expand;
+  auto grad_t = loss_grad_expand;
```
Comment (zhwesky2010) [inline paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h:68]: 这个修改不影响计算结果吗？

File: paddle/phi/kernels/impl/kldiv_loss_kernel_impl.h [hunk 3]
```diff
@@ -24,21 +24,29 @@ namespace phi {
...
 struct KLDivLossForward {
-  HOSTDEVICE KLDivLossForward() {}
+  bool LogTarget = false;
+
+  HOSTDEVICE KLDivLossForward(bool _log_target) : LogTarget(_log_target) {}
 
   HOSTDEVICE T operator()(const T& target, const T& input) const {
-    if (target <= 0) {
-      return 0;
+    if (LogTarget) {
```
Comment (zhwesky2010) [inline paddle/phi/kernels/impl/kldiv_loss_kernel_impl.h:32]: 同上

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -1702,8 +1709,12 @@ def kl_div(input, label, reduction='mean', name=None):
     ):
         label = paddle.cast(label, 'float64')
 
+    if paddle.is_compiled_with_xpu():
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py:1712]: 这个是需要修改xpu 的kernel，你学一学，也不难，主要是调一些xpu::的数学库函数可以实现exp等操作

**Review by zhwesky2010 — COMMENTED** [review]
将两个PR拆开吧，一个是修改max_unpool1d/2d/3d，一个是修改kl_div

File: paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h (hunk 1 above)
Comment (NKNaN) [reply paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h:28]: 好的

File: paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h (hunk 2 above)
Comment (NKNaN) [reply paddle/phi/kernels/impl/kldiv_loss_grad_kernel_impl.h:68]: 不会影响计算结果，target_t 我放到 KLDivLossBackward 这个结构体去乘了

之前是
```cpp
      return static_cast<T>(-1.) * grad;
```

改为
```cpp
      return static_cast<T>(-1.) * target * grad;
```

验证了一下是不影响的

File: paddle/phi/kernels/impl/kldiv_loss_kernel_impl.h (hunk 3 above)
Comment (NKNaN) [reply paddle/phi/kernels/impl/kldiv_loss_kernel_impl.h:32]: 好的

File: python/paddle/nn/functional/loss.py (hunk 4 above)
Comment (NKNaN) [reply python/paddle/nn/functional/loss.py:1712]: 好的，这个 xpu::kldiv_loss 他的定义是在哪里
image

Comment (zhwesky2010) [reply python/paddle/nn/functional/loss.py:1712]: > 好的，这个 xpu::kldiv_loss 他的定义是在哪里 image

`xpu::` 这种开头的就是xpu自己提供的sdk了，可以理解为其他厂商提供的对外硬件接口，类似于CUDA库的cuda::函数，你需要查一查这些接口是否提供相应的功能，不提供的话就需要额外组装来计算
